File: src/main/java/org/willxu/algorithm/service/string/ExcelSheetColumnTitle.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.string;

public interface ExcelSheetColumnTitle {
    /**
     * Given an integer columnNumber, return its corresponding column
     * title as it appears in an Excel sheet.
     *
     * For example:
     * A -> 1
     * B -> 2
     * c -> 3
     * ...
     * z -> 26
     * AA -> 27
     * AB -> 28
     * ...
     * @param columnNumber 1 <= columnNumber <= 2^31 - 1
     * @return excel sheet column title code
     */
    String convertToTitle(int columnNumber);
}

File: src/leetcode/p0100_0199/p0168_excel_sheet_column_title/excel_sheet_column_title.py
from abc import ABC, abstractmethod

class ExcelSheetColumnTitle(ABC):
    """
    Given an integer columnNumber, return its corresponding column title as it appears in an Excel sheet.
    """
    @abstractmethod
    def convertToTitle(self, columnNumber: int) -> str:
        """
        Args:
            columnNumber:
                1 <= columnNumber <= 2^31 - 1
        """
        pass

class ExcelSheetColumnTitleLoop(ExcelSheetColumnTitle):
    def convertToTitle(self, columnNumber: int) -> str:
        result = ""
        while columnNumber > 0:
            columnNumber -= 1
            result = chr(columnNumber % 26 + ord('A')) + result
            columnNumber //= 26
        return result
